arch/arm/arm64/include/uk/asm/spinlock.h

```diff
@@ -0,0 +1,117 @@
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __UKARCH_SPINLOCK_H__
+#error Do not include this header directly
+#endif
+
+#include <uk/arch/atomic.h>
+
+/* Unless you know what you are doing, use uk_spinlock instead. */
+typedef struct __spinlock {
+	volatile int lock;
+} __spinlock;
+
+/* Initialize a spinlock to unlocked state */
+#define UKARCH_SPINLOCK_INITIALIZER() { 0 }
+
+/**
+ * Initialize a spinlock to unlocked state.
+ * @param [in/out] lock Pointer to spinlock
+ */
+static inline void ukarch_spin_init(__spinlock *lock)
+{
+	lock->lock = 0;
+}
+
+/**
+ * Acquire lock. It is guaranteed that the spinlock will be held exclusively.
+ * @param [in/out] lock Pointer to spinlock
+ */
+static inline void ukarch_spin_lock(__spinlock *lock)
+{
+	register int r, locked = 1;
+
+	__asm__ __volatile__(
+		"	sevl\n"			/* set event locally */
+		"1:	wfe\n"			/* wait for event */
+		"2:	ldaxr	%w0, [%1]\n"	/* exclusive load lock value */
+		"	cbnz	%w0, 1b\n"	/* check if already locked */
+		"	stxr	%w0, %2, [%1]\n"/* try to lock it */
+		"	cbnz	%w0, 2b\n"	/* jump to l2 if we failed */
+		: "=&r" (r)
+		: "r" (&lock->lock), "r" (locked));
+}
+
+/**
+ * Release previously acquired lock.
+ * @param [in/out] lock Pointer to spinlock
+ */
+static inline void ukarch_spin_unlock(__spinlock *lock)
+{
+	__asm__ __volatile__(
+		"stlr	wzr, [%0]\n"		/* unlock lock */
+		"sev\n"				/* wake up any waiters */
+		:
+		: "r" (&lock->lock));
+}
+
+/**
+ * Try to acquire lock. If the lock is already acquired (busy), this function
+ * returns instead of spinning.
+ * @param [in/out] lock Pointer to spinlock
+ * @return A non-zero value if lock was acquired, 0 otherwise
+ */
+static inline int ukarch_spin_trylock(__spinlock *lock)
+{
+	register int r, locked = 1;
+
+	__asm__ __volatile__(
+		"	ldaxr	%w0, [%1]\n"	/* exclusive load lock value */
+		"	cbnz	%w0, 1f\n"	/* bail out if locked */
+		"	stxr	%w0, %2, [%1]\n"/* try to lock it */
+		"1:\n"
+		: "=&r" (r)
+		: "r" (&lock->lock), "r" (locked));
+
+	return r;
```

Comment:
This actually returns zero on successful acquisition, but the documentation states otherwise. Btw, I found this interface inconsistent with is_locked(). Maybe we should leave the return value as is and change the interface instead?

Reply:
```suggestion
	return !r;
```

I missed this one 👍🏻 Very good that you are testing everything!